File: src/main/java/me/kuku/yuq/controller/WeiboController.java
package me.kuku.yuq.controller;

import com.IceCreamQAQ.Yu.annotation.Action;
import com.IceCreamQAQ.Yu.annotation.Before;
import com.IceCreamQAQ.Yu.job.JobManager;
import com.icecreamqaq.yuq.annotation.GroupController;
import com.icecreamqaq.yuq.annotation.QMsg;
import com.icecreamqaq.yuq.entity.Group;
import com.icecreamqaq.yuq.message.MessageItemFactory;
import me.kuku.pojo.Result;
import me.kuku.yuq.entity.QqEntity;
import me.kuku.yuq.entity.WeiboEntity;
import me.kuku.yuq.entity.WeiboService;
import me.kuku.yuq.logic.WeiboLogic;
import me.kuku.yuq.pojo.WeiboPojo;

import javax.inject.Inject;
import java.io.IOException;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

@GroupController
public class WeiboController {

	@Inject
	private WeiboLogic weiboLogic;
	@Inject
	private WeiboService weiboService;
	@Inject
	private MessageItemFactory mif;
	@Inject
	private JobManager jobManager;

	@Action("热搜")
	@QMsg(at = true, atNewLine = true)
	public String hotSearch() throws IOException {
		List<String> list = weiboLogic.hotSearch();
		StringBuilder sb = new StringBuilder();
		for (String str: list){
			sb.append(str).append("\n");
		}
		return sb.deleteCharAt(sb.length() - 1).toString();
	}

	@Action("微博用户 {username}")
	@QMsg(at = true, atNewLine = true)
	public String wbInfo(String username) throws IOException {
		Result<List<WeiboPojo>> idResult = weiboLogic.getIdByName(username);
		List<WeiboPojo> idList = idResult.getData();
		if (idList == null) return idResult.getMessage();
		return weiboLogic.getUserInfo(idList.get(0).getUserId());
	}

	@Action("微博二维码")
	public void qrcode(long qq, QqEntity qqEntity, Group group) throws IOException {
		Map<String, String> map = weiboLogic.loginByQr1();
		group.sendMessage(mif.at(qq).plus("请使用微博APP扫码登录").plus(mif.imageByUrl("https:" + map.get("url"))));
		jobManager.registerTimer(() -> {
			String id = map.get("id");
			while (true){
				try {
					TimeUnit.SECONDS.sleep(3);
					Result<WeiboEntity> result = weiboLogic.loginByQr2(id);
					if (result.getCode() == 200){
						WeiboEntity newWeiboEntity = result.getData();
						WeiboEntity weiboEntity = weiboService.findByQqEntity(qqEntity);
						if (weiboEntity == null) weiboEntity = WeiboEntity.Companion.getInstance(qqEntity);
						weiboEntity.setPcCookie(newWeiboEntity.getPcCookie());
						weiboEntity.setMobileCookie(newWeiboEntity.getMobileCookie());
						weiboService.save(weiboEntity);
						group.sendMessage(mif.at(qq).plus("绑定微博成功！"));
						break;
					}else if (result.getCode() == 500){
						group.sendMessage(mif.at(qq).plus(result.getMessage()));
						break;
					}
				}catch (Exception e){
					e.printStackTrace();
					break;
				}
			}
		}, 0);
	}

	@Before(only = "weiboMonitor")
	@QMsg(at = true)
	public WeiboEntity before(QqEntity qqEntity){
		WeiboEntity weiboEntity = weiboService.findByQqEntity(qqEntity);
		if (weiboEntity == null)
			throw mif.at(qqEntity.getQq()).plus("您还没有绑定微博，请发送<微博二维码>绑定账号").toThrowable();
		else return weiboEntity;
	}

	@Action("微博监控 {status}")
	@QMsg(at = true)
	public String weiboMonitor(boolean status, WeiboEntity weiboEntity){
		weiboEntity.setMonitor(status);
		weiboService.save(weiboEntity);
		return status ? "微博监控开启成功！" : "微博监控关闭成功！";
	}

}

File: src/main/kotlin/me/kuku/yuq/controller/WeiboController.kt
package me.kuku.yuq.controller

import com.IceCreamQAQ.Yu.annotation.Action
import com.IceCreamQAQ.Yu.annotation.Before
import com.icecreamqaq.yuq.annotation.GroupController
import com.icecreamqaq.yuq.annotation.PrivateController
import com.icecreamqaq.yuq.controller.BotActionContext
import com.icecreamqaq.yuq.mif
import kotlinx.coroutines.delay
import me.kuku.utils.OkHttpUtils
import me.kuku.yuq.entity.QqEntity
import me.kuku.yuq.entity.WeiboEntity
import me.kuku.yuq.entity.WeiboService
import me.kuku.yuq.logic.WeiboLogic
import org.springframework.stereotype.Component

@GroupController
@PrivateController
@Component
class WeiboController (
    private val weiboService: WeiboService
) {

    @Action("微博登录")
    suspend fun login(qqEntity: QqEntity, context: BotActionContext) {
        val weiboQrcode = WeiboLogic.loginByQr1()
        val url = weiboQrcode.url
        val bytes = OkHttpUtils.getBytes("https:$url")
        context.source.sendMessage(mif.at(qqEntity.qq).plus(mif.imageByByteArray(bytes)).plus("请使用微博APP扫码登录（第三方微博也可以）").toMessage())
        while (true) {
            delay(3000)
            val result = WeiboLogic.loginByQr2(weiboQrcode)
            if (result.success()) {
                val newWeiboEntity = result.data()
                val weiboEntity = weiboService.findByQqEntity(qqEntity) ?: WeiboEntity().also {
                    it.qqEntity = qqEntity
                }
                weiboEntity.pcCookie = newWeiboEntity.pcCookie
                weiboEntity.mobileCookie = newWeiboEntity.mobileCookie
                weiboService.save(weiboEntity)
                context.source.sendMessage(mif.at(qqEntity.qq).plus("绑定微博成功"))
                break
            } else if (result.code in listOf(201, 202)) continue
            else {
                context.source.sendMessage(mif.at(qqEntity.qq).plus(result.message))
                break
            }
        }
    }

    @Before(except = ["login"])
    fun before(qqEntity: QqEntity): WeiboEntity {
        return weiboService.findByQqEntity(qqEntity) ?: throw mif.at(qqEntity.qq).plus("您未绑定微博账号，操作失败").toThrowable()
    }

    @Action("微博超话签到")
    suspend fun superSign(weiboEntity: WeiboEntity): String {
        val result = WeiboLogic.superTalkSign(weiboEntity)
        return if (result.success()) "微博超话签到成功"
        else "微博超话签到失败，${result.message}"
    }

    @Action("微博推送 {status}")
    fun weiboPush(status: Boolean, weiboEntity: WeiboEntity): String {
        weiboEntity.config.push = status.toStatus()
        weiboService.save(weiboEntity)
        return "微博推送${if (status) "开启" else "关闭"}成功"
    }

    @Action("微博自动签到 {status}")
    fun weiboAutoSign(status: Boolean, weiboEntity: WeiboEntity): String {
        weiboEntity.config.sign = status.toStatus()
        weiboService.save(weiboEntity)
        return "微博自动签到${if (status) "开启" else "关闭"}成功"
    }


}